search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt

```diff
@@ -0,0 +1,121 @@
+        return if (statusCode.is2xxSuccessful) {
+            logger.info("Successfully received Informations from CSR at : $uri")
+
+            response
+        } else {
+            logger.info("Error contacting CSR at : $uri")
+            logger.info("Error contacting CSR at : $response")
+            null
+        }
+    }
+
+    fun mergeEntity(
+        localEntity: CompactedEntity?,
+        pairsOfEntitiyWithISAuxiliary: List<CompactedEntityWithIsAuxiliary>
+    ): CompactedEntity? {
+        if (localEntity == null && pairsOfEntitiyWithISAuxiliary.isEmpty()) return null
+
+        val mergedEntity = localEntity?.toMutableMap() ?: mutableMapOf()
+
+        pairsOfEntitiyWithISAuxiliary.forEach {
+                entity ->
+            entity.first.entries.forEach {
+                    (key, value) ->
+                when {
+                    !mergedEntity.containsKey(key) -> mergedEntity[key] = value
+                    JSONLD_COMPACTED_ENTITY_CORE_MEMBERS.contains(key) -> {}
+                    else -> mergedEntity[key] = mergeAttribute(mergedEntity[key]!!, value, entity.second)
+                }
+            }
+        }
+        return mergedEntity
+    }
+
+    /**
+     * Implements 4.5.5 - Multi-Attribute Support
+     */
+    private fun mergeAttribute(
+        attribute1: Any,
+        attribute2: Any,
+        auxiliary: Boolean = false
+    ): CompactedAttribute {
+        val mergeMap = attributeToDatasetIdMap(attribute1).toMutableMap()
+        val attribute2Map = attributeToDatasetIdMap(attribute2)
+        attribute2Map.entries.forEach { (datasetId, value) ->
+            when {
+                mergeMap[datasetId] == null -> mergeMap[datasetId] = value
+                auxiliary -> {}
+                mergeMap[datasetId]!!.isBefore(value, NGSILD_OBSERVED_AT_TERM) -> mergeMap[datasetId] = value
+                value.isBefore(mergeMap[datasetId]!!, NGSILD_OBSERVED_AT_TERM) -> {}
+                mergeMap[datasetId]!!.isBefore(value, NGSILD_MODIFIED_AT_TERM) -> mergeMap[datasetId] = value
+                value.isBefore(mergeMap[datasetId]!!, NGSILD_MODIFIED_AT_TERM) -> {}
+                nextBoolean() -> mergeMap[datasetId] = value
+                else -> {}
+            }
+        }
+        return mergeMap.values.toList()
+    }
+
+    private fun attributeToDatasetIdMap(attribute: Any): Map<String?, Map<String, Any>> = when (attribute) {
+        is Map<*, *> -> {
+            attribute as SingleAttribute
+            mapOf(attribute[NGSILD_DATASET_ID_TERM] as? String to attribute)
+        }
+        is List<*> -> {
+            attribute as CompactedAttribute
+            attribute.associate {
+                it[NGSILD_DATASET_ID_TERM] as? String to it
+            }
+        }
+
+        else -> throw InternalErrorException(
+            "the attribute is nor a list nor a map, check that you have excluded the CORE Members"
+        )
```

Comment:
please don't throw exceptions 🙏

Reply:
not supposed to happen does not mean it will never happen. and throwing an exception will stop the request and the end user will receive a 500. not the expected behavior?

Reply:
It can't happen as long as the function is call for non CORE attribute. (which is his purpose) 
If the error happen that means its a development issue.
I could replace it with : 
`        is Map<*, *> -> {
            attribute as CompactedAttributeInstance
            mapOf(attribute[NGSILD_DATASET_ID_TERM] as? String to attribute)
        }
        else -> {
            attribute as CompactedAttributeInstances
            attribute.associateBy { it[NGSILD_DATASET_ID_TERM] as? String }
        }`
        to avoid calling explicitely the throw if you prefer. (but it will have the same behavior)

Reply:
Or it is a context source returning an invalid payload (and should be translated to a 111 - Revalidation Failed)